sklearn/preprocessing/_encoders.py

```diff
@@ -87,9 +90,17 @@ def _fit(self, X, handle_unknown='error'):
             else:
                 cats = np.array(self.categories[i], dtype=Xi.dtype)
                 if Xi.dtype != object:
-                    if not np.all(np.sort(cats) == cats):
-                        raise ValueError("Unsorted categories are not "
-                                         "supported for numerical categories")
+                    sorted_cats = np.sort(cats)
+                    error_msg = ("Unsorted categories are not "
+                                 "supported for numerical categories")
+                    # if there are nans, nan should be the last element
+                    if np.isnan(sorted_cats[-1]):
+                        if not (np.isnan(cats[-1]) or
+                                np.all(sorted_cats[:-1] == cats[:-1])):
+                            raise ValueError(error_msg)
+                    else:
+                        if not np.all(sorted_cats == cats):
+                            raise ValueError(error_msg)
```

Comment:
I find this logic hard to read. Is the following better?
```suggestion
                    stop_idx = -1 if np.isnan(sorted_cats[-1]) else None
                    if (not np.all(sorted_cats[:stop_idx] == cats[:stop_idx]) or
                            (np.isnan(sorted_cats[-1]) and not np.isnan(cats[-1]))):
                        raise ValueError(error_msg)
```

Reply:
I don't think the `np.isnan(sorted_cats[-1]) and not np.isnan(cats[-1])` is tested